activerecord/lib/active_record/connection_adapters/metadata_registry.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+require "digest/md5"
+
+module ActiveRecord
+  module ConnectionAdapters # :nodoc:
+    module MetadataRegistry
+      REGISTRY = {}
+
+      class << self
+        def fetch(type, *args)
+          key = Digest::MD5.hexdigest(Marshal.dump([type.name, args]))
+          REGISTRY[key] ||= yield.freeze
+        end
+      end
+
+      module Concern
+        def new(*args)
```

Comment:
Why should deduplication be `new`'s responsibility?

Reply:
I initially implemented this as a monkey patch, so it was much easier that way. If you feel stringly against this I can see about using an alternative constructor.

I also have to make sure the hook will be used by YAML, because that is what is used to serialize and deserialize the schema cache.

Reply:
I'm just thinking out loud about where the hook for this should be and where the boundaries of the deduplication should go.

Reply:
Yup. YAML don't go through `new` anyway, so that won't be enough.

Reply:
Oh right, it goes through `init_with` 👍

Reply:
Yup. I just went into the source to confirm, there's just no way to deduplicate data with YAML. So I'll have to do the interning after the schema is deserialized, which call for some `-@` method I guess.